llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -13678,9 +13687,88 @@ static SDValue performSETCCCombine(SDNode *N, SelectionDAG &DAG,
                                    const RISCVSubtarget &Subtarget) {
   SDValue N0 = N->getOperand(0);
   SDValue N1 = N->getOperand(1);
+  ISD::CondCode Cond = cast<CondCodeSDNode>(N->getOperand(2))->get();
   EVT VT = N->getValueType(0);
   EVT OpVT = N0.getValueType();
+  SDLoc DL(N);
+
+  // Both rules are looking for an equality compare.
+  if (!isIntEqualitySetCC(Cond))
+    return SDValue();
+
+  // Rule 1
+  using namespace SDPatternMatch;
+  auto matchSelectCC = [](SDValue Op, SDValue VLCandidate, bool Inverse,
+                          SDValue &Select) -> bool {
+    // Remove any sext or zext
+    auto ExtPattern =
+        m_AnyOf(m_Opc(ISD::SIGN_EXTEND_INREG), m_And(m_Value(), m_AllOnes()));
```

Comment:
m_And(m_Value(), m_AllOnes()) isn't a zero extend and should be removed by DAG combine

Reply:
I'm not sure m_AllOnes is implemented correctly. It uses `APInt::isSameValue` which zero extends when bit widths don't match. `m_AllOnes` should be checking that all bits are one. If `m_AllOnes` is used to compare against a value that is less than 64 bits, that value will be zero extended to 64 bits and will fail to match.

Reply:
It's fixed now. In this particular case I think VL will almost certain be zext from i32 so I'm pattern matching `(and X, <32 trailing ones>)` here instead.